File: src/test/SequentialReaderTest.cpp
//
// Copyright 2020, Jefferson Science Associates, LLC.
// Subject to the terms in the LICENSE file found in the top-level directory.
//
// EPSCI Group
// Thomas Jefferson National Accelerator Facility
// 12000, Jefferson Ave, Newport News, VA 23606
// (757)-269-7100


#include <iostream>
#include <fstream>

#include "TestBase.h"
#include "eviocc.h"


namespace evio {


    /**
     * Test swapping evio data.
     *
     * @date 03/17/2025
     * @author timmer
     */
    class SequentialReaderTest : public TestBase {
    };
}

    int main(int argc, char **argv) {

        try {
            int counter = 1;

            // Write 3 events to file
            std::string fileName  = "/tmp/seqReadTest.evio";
            evio::SequentialReaderTest tester;
            auto ev = tester.createEventBuilderEvent(1,1);
            evio::EventWriter writer(fileName);
            writer.writeEvent(ev);
            writer.writeEvent(ev);
            writer.writeEvent(ev);
            writer.close();

            std::cout << "read ev file: " << fileName << std::endl;


            evio::EvioReader fileReader(fileName, false, true);

            std::cout << "count events ..." << std::endl;
            size_t eventCount = fileReader.getEventCount();
            std::cout << "done counting events, " << eventCount << std::endl;

            // Read sequentially

            while (fileReader.parseNextEvent() != nullptr) {
                std::cout <<  "parseNextEvent # " << counter++ << std::endl;
            }

            // Now read non-sequentially

            std::cout << "get ev #1" << std::endl;
            auto event = fileReader.getEvent(1);

            std::cout << "get ev #2" << std::endl;
            event = fileReader.getEvent(2);

            std::cout << "get ev #3" << std::endl;
            event = fileReader.getEvent(3);



            std::cout << "goto ev #1" << std::endl;
            event = fileReader.gotoEventNumber(1);

            std::cout << "goto ev #2" << std::endl;
            event = fileReader.gotoEventNumber(2);

            std::cout << "goto ev #3" << std::endl;
            event = fileReader.gotoEventNumber(3);



            std::cout << "parse ev #1" << std::endl;
            event = fileReader.parseEvent(1);

            std::cout << "parse ev #2" << std::endl;
            event = fileReader.parseEvent(2);

            std::cout << "parse ev #3" << std::endl;
            event = fileReader.parseEvent(3);


            // Rewind
            std::cout << "rewind file" << std::endl;
            fileReader.rewind();


            // Read sequentially again
            counter = 1;
            while (fileReader.parseNextEvent() != nullptr) {
                std::cout << "parseNextEvent # " << counter++ << std::endl;
            }

        }
        catch (evio::EvioException & e) {
            std::cerr << "Error: " << e.what() << std::endl;
        }
    }

File: etc/src/test/java/jevio/SequentialReaderTest.java
package org.jlab.coda.jevio.test;

import org.jlab.coda.jevio.*;

import java.io.File;


/**
 * Test program.
 * @author timmer
 * Date: Apr 14, 2015
 */
public class SequentialReaderTest extends TestBase {

    /** For testing only */
    public static void main(String args[]) {


        try {
            int counter = 1;

            // Write 3 events to file
            String fileName  = "/tmp/seqReadTest.evio";
            SequentialReaderTest tester = new SequentialReaderTest();
            EvioEvent ev = tester.createEventBuilderEvent(1,1);
            EventWriter writer = new EventWriter(fileName);
            writer.writeEvent(ev);
            writer.writeEvent(ev);
            writer.writeEvent(ev);
            writer.close();

            File fileIn = new File(fileName);
            System.out.println("read ev file: " + fileName + " size: " + fileIn.length());


            EvioReader fileReader = new EvioReader(fileName, false, true);

            System.out.println("count events ...");
            int eventCount = fileReader.getEventCount();
            System.out.println("done counting events, " + eventCount);

            // Read sequentially

            while (fileReader.parseNextEvent() != null) {
                System.out.println("parseNextEvent # " + counter++);
            }

            // Now read non-sequentially

            System.out.println("get ev #1");
            EvioEvent event = fileReader.getEvent(1);

            System.out.println("get ev #2");
            event = fileReader.getEvent(2);

            System.out.println("get ev #3");
            event = fileReader.getEvent(3);



            System.out.println("goto ev #1");
            event = fileReader.gotoEventNumber(1);

            System.out.println("goto ev #2");
            event = fileReader.gotoEventNumber(2);

            System.out.println("goto ev #3");
            event = fileReader.gotoEventNumber(3);



            System.out.println("parse ev #1");
            event = fileReader.parseEvent(1);

            System.out.println("parse ev #2");
            event = fileReader.parseEvent(2);

            System.out.println("parse ev #3");
            event = fileReader.parseEvent(3);


            // Rewind
            System.out.println("rewind file");
            fileReader.rewind();


            // Read sequentially again
            counter = 1;
            while (fileReader.parseNextEvent() != null) {
                System.out.println("parseNextEvent # " + counter++);
            }

        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }


}
